interfaces: update cups-control and add cups for providing snaps

jdstrand opened the pull request:
* interfaces/cups-control: implement cups-control provided by snap
* interfaces: add cups companion interface to cups-control
    
    On systems where the slot is provided by an app snap, the cups interface is
    the companion interface to the cups-control interface. The design of these
    interfaces is based on the idea that the slot implementation (eg cupsd) is
    expected to query snapd on if the cups-control interface is connected or not
    and the print service will mediate admin functionality (ie, the rules in
    these interfaces allow connecting to the print service, but do not implement
    enforcement rules; it is up to the print service to provide enforcement).
    
    For now we limit this to app-providing slot implementations since there
    are currently no classic systems with a mediating cupsd. We can revisit
    this when it becomes available.

References:
* https://forum.snapcraft.io/t/interface-request-cups-control-on-cups-snap-and-including-d-bus/15233/9

jdstrand commented:
Marked as blocked for now since it isn't ready for review.

jdstrand commented:
This is now ready for review.

anonymouse64 commented on interfaces/builtin/cups.go:
```diff
@@ -0,0 +1,61 @@
+// On systems where the slot is provided by an app snap, the cups interface is
+// the companion interface to the cups-control interface. The design of these
+// interfaces is based on the idea that the slot implementation (eg cupsd) is
+// expected to query snapd on if the cups-control interface is connected or not
+// and the print service will mediate admin functionality (ie, the rules in
```
```suggestion
// expected to query snapd to determine if the cups-control interface is connected or not for the cups client
// and the print service will mediate admin functionality (ie, the rules in
```

jdstrand replied:
@anonymouse64 - I went with: `expected to query snapd to determine if the cups-control interface is connected or not for the peer client process and the ...`

jdstrand commented:
FYI, Til from the desktop commented on his testing here: https://discourse.ubuntu.com/t/desktop-team-updates-monday-29th-june-2020/17025/7

anonymouse64 commented on interfaces/builtin/cups_control_test.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func (s *cupsControlSuite) TearDownTest(c *C) {
+	dirs.SetRootDir("/")
```
same comment as the other PR, this could be done with `testutil.BaseTest` and `AddCleanup` but this is a minor nitpick

jdstrand replied:
Actually, TearDown isn't needed at all since we aren't calling dirs.SetRootDir on a temp dir or anything, so I've removed this.

anonymouse64 commented on interfaces/builtin/cups_control_test.go:
```diff
@@ -0,0 +1,188 @@
+	c.Assert(interfaces.BeforePreparePlug(s.iface, s.plugInfo), IsNil)
+}
+
+func (s *cupsControlSuite) TestAppArmorSpecCore(c *C) {
```
thanks for the extensive test!

anonymouse64 commented on interfaces/builtin/cups_test.go:
```diff
@@ -0,0 +1,115 @@
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "#include <abstractions/cups-client>")
+}
+
+func (s *cupsSuite) TestAppArmorSpecClassic(c *C) {
```
minor nitpick, but since the behavior is the same on classic and core you could collapse the two cases into a single test with
```go
for _, onClassic := range []bool{true,false} {
    restore := release.MockOnClassic(onClassic)
    ...
    restore()
}
```

anonymouse64 commented on tests/main/interfaces-cups/task.yaml:
```diff
@@ -0,0 +1,66 @@
+    server.
+
+prepare: |
+    snap pack test-snapd-provider
+    snap install --dangerous ./test-snapd-provider_1_all.snap
```
minor nitpick, I think this can be simplified to just `install_local test-snapd-provider`

anonymouse64 commented on tests/main/interfaces-cups/task.yaml:
```diff
@@ -0,0 +1,66 @@
+prepare: |
+    snap pack test-snapd-provider
+    snap install --dangerous ./test-snapd-provider_1_all.snap
+    snap pack test-snapd-consumer
+    snap install --dangerous ./test-snapd-consumer_1_all.snap
```
same

anonymouse64 commented on tests/main/interfaces-cups/task.yaml:
```diff
@@ -0,0 +1,66 @@
+    snap install --dangerous ./test-snapd-consumer_1_all.snap
+
+    if [ -e /run/cups ]; then
+        mv /run/cups /run/cups.orig
```
fun trick, in master you can now do this in prepare:

```bash
tests.cleanup prepare
tests.cleanup defer mv /run/cups.orig /run/cups
```

and then in restore do

```bash
tests.cleanup restore
```

but I won't block on this

zyga replied:
We also have fs-state helpers that handle the backup/restore case directly. Both can be used together defer restore

anonymouse64 commented on tests/main/interfaces-cups/test-snapd-consumer/bin/sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+PS1='$ '
```
what's this for here? just to aid in debugging?

jdstrand replied:
Many of the `tests/main/interfaces-*` tests have this and I just copied it over.

zyga replied:
This prevents a bug in sh, which cannot cope with PS1 set by spread, causing an infinite loop. It allows running `snap run test-snapd-consumer.sh` without an error

anonymouse64 replied:
@zyga should this be in every snap that runs `/bin/sh` as a snap app?

anonymouse64 commented on tests/main/interfaces-cups/test-snapd-consumer/meta/snap.yaml:
```diff
@@ -0,0 +1,12 @@
+version: 1
+architectures: [all]
+apps:
+  cups-control:
+    command: bin/sh
+    plugs:
+    - cups-control
+  cups:
+    command: bin/sh
+    plugs:
+    - cups
```
I like that there is an app for each plug, it makes the test a good bit simpler, thanks for this!

anonymouse64 commented on tests/main/interfaces-cups/task.yaml:
```diff
@@ -0,0 +1,66 @@
+summary: Ensure that the cups interfaces work with app providers
+
```
oh we probably need a `systems: [ubuntu-*]` or something for the platforms that have strict confinement, the test as written fails on systems without strict confinement:

```
2020-07-02T18:40:21.1169826Z ##[error]2020-07-02 18:40:21 Error executing google:fedora-32-64:tests/main/interfaces-cups (jul021808-861954) : 
2020-07-02T18:40:21.1179219Z -----
2020-07-02T18:40:21.1180102Z + test-snapd-provider.sh -c 'echo slot > /run/cups/cups.sock'
2020-07-02T18:40:21.1180405Z WARNING: cgroup v2 is not fully supported yet, proceeding with partial confinement
2020-07-02T18:40:21.1181129Z + test-snapd-provider.sh -c 'echo slot > /run/cups/other'
2020-07-02T18:40:21.1181408Z WARNING: cgroup v2 is not fully supported yet, proceeding with partial confinement
2020-07-02T18:40:21.1182151Z + not test-snapd-consumer.cups-control -c 'head /run/cups/cups.sock'
2020-07-02T18:40:21.1183071Z WARNING: cgroup v2 is not fully supported yet, proceeding with partial confinement
2020-07-02T18:40:21.1183344Z slot
2020-07-02T18:40:21.1183992Z -----
```

pedronis commented on interfaces/builtin/cups_control.go:
```diff
@@ -19,30 +19,159 @@
+        - app
         - core
     deny-auto-connection: true
+    deny-connection:
+      on-classic: false
```
so to clarify for everybody, this interface was also before implicit on classic, it wasn't and isn't implicit on core. Now app-provided slots can exist for it.

Before manual connection was allowed on classic and core but on core no slot could actually exist.
Now slots can exist on both core and classic,  on classic provided by system or app, on core provided by app.

This allows manual connection on classic to either I think, but stops it on core...

* it needs a test I think in base_declaration_test.go
* another possibility would be to prohibit in both places manual connection if the slot is from an app?  (we don't seem ATM to allow the syntax for that but we could... (we support only plug-snap-types in slot connection constraints, but we could support slot-snap-types as well in theory)

jdstrand replied:
Note, for the base declaration changes to cups-control, I followed the pattern we used for avahi-observe/avahi-control since it is the same: it can be implicit on classic or app installed on core and classic. It cannot be implicit on core. (This is in contrast to the new cups interface which today can only be via app snap).

@anonymouse64, this is why we made the distinction here but you are right that it is somewhat awkward. What we are trying to achieve is not requiring a connection constraint via snap declaration for an implicit slot, and this does that. It does a little more though since it grants connection to an app snap on classic. This is 'ok' because the review-tools understand this established pattern and will flag for manual review a snap that is trying to slots cups-control. We might want to do something like this:

```
const cupsControlBaseDeclarationSlots = ` 
  cups-control:
    allow-installation:
      slot-snap-type:
        - app
        - core
    deny-auto-connection: true
    allow-connection:
      on-classic: true
      slot-snap-type:
      - core
```

but as @pedronis mentioned, `slot-snap-type` is not allowed in a connection constraint for `slots`. `slot-snap-type` is allowed in the connection constraint for `plugs` though, so we could use this in the base declaration:

```
const cupsControlBaseDeclarationSlots = ` 
  cups-control:
    allow-installation:
      slot-snap-type:
        - app
        - core

const cupsControlBaseDeclarationPlugs = `
  cups-control:
    allow-installation:
      plug-snap-type:
        - app
    deny-connection:
      on-classic: true
      slot-snap-type:
        - app
    deny-auto-connection: true
```

which is more accurate, but it falls outside of our established patterns and is arguably more difficult to understand (note, the review-tools would still flag something slotting cups-control with the above for manual review). IME, if we determine that we want a new pattern for things like avahi and cups, then those base declaration changes should be in a followup PR. I think I would prefer adding 'slot-snap-types' to the connection constraint instead of contorting the existing syntax like this though.

@pedronis - the tests were updated in the same manner as for avahi-control/avahi-observe by adding `cups-control` to `slotInstallation` in basedeclaration_test.go. I'll take a look at adding more tests for this pattern (avahi* and cups-control).

pedronis replied:
I unresolved this convo because otherwise the link added in the tests comment doesn't get here

jdstrand commented on interfaces/builtin/cups.go:
```diff
@@ -0,0 +1,61 @@
+// design allows the snap provider to slots both cups-control and cups or just
+// cups-control (like with implicit classic or any slot provider without
+// mediation patches), but not just cups).
+const cupsBaseDeclarationSlots = `
```
> not necessarily related to this PR, but is there any plan to make auto-connect by snaps that `plugs: [cups]` to auto-connect automatically to the "cups" providing snap? or are we not going to enable that, and just allow snaps to request (and grant) auto-connect for cups interface plugs to the cups snap cups slot (that's a lot of coffee smile )

Whether or not we auto-connect the cups interface will be entirely dependent on the slot provider which is why the base declaration is set the way it is, which uses the common pattern that by default, connect and auto-connect are disabled. If we do the vetting, etc, etc in forum requests for a slot provider, we'll grant connection. If it has the mediation patches, we'll grant auto-connection. The addition of the cups interface in this PR is in response to Till's work to add and upstream mediation patches to CUPS (so it is expected that the CUPS snap the slots 'cups' would allow auto-connect).

If those patches are backported to the CUPS debs in Ubuntu, we might consider auto-connection for implicit classic, like we do today with audio-playback (since most distros that support strict are forks/derivatives that pull from the Ubuntu archive and would therefore receive the mediation patches as well), but we can take a look at that if/when the mediation patches are SRU'd.

jdstrand commented on tests/main/interfaces-cups/task.yaml:
```diff
@@ -0,0 +1,66 @@
+
+    # The interface can be disconnected
+    snap disconnect test-snapd-consumer:cups-control
+    not test-snapd-consumer.cups-control -c "head /run/cups/cups.sock"
```
It would, yes. Nice catch.

jdstrand replied:
Done

jdstrand commented on interfaces/builtin/cups.go:
```diff
@@ -0,0 +1,62 @@
+// On systems where the slot is provided by an app snap, the cups interface is
+// the companion interface to the cups-control interface. The design of these
+// interfaces is based on the idea that the slot implementation (eg cupsd) is
+// expected to query snapd to determine if the cups-control interface is
```
> should that be "if the cups interface" here?

Actually, no. The idea is that since the socket is allowed in both interfaces, cupsd will see if 'cups-control' is connected to determine if admin functionality is allowed (cupsd can assume that since communications are happening at all that one of them is connected, so if cups-control isn't connected, only allow printing and not admin).

jdstrand commented on interfaces/builtin/cups_control.go:
```diff
@@ -19,30 +19,159 @@
+
+	// implement 'implicitOnCore: false/implicitOnClassic: true' by only
+	// applying the snippet if the slot is an app or we are on classic
+	if slot.Snap().Type() == snap.TypeApp || release.OnClassic {
```
@anonymouse64 - when I was developing the test cases and looked at avahi-control/observe, I discovered that we didn't have complete coverage and there was nothing in AppArmorConnectedPlug() that was making sure 'implicitOnCore: false/implicitOnClassic: true' was honored. Eg, if I comment out this conditional, then `// core to consumer on core is empty for ConnectedPlug` fails in `TestAppArmorSpecCore`, which is completely expected since we are calling AddConnectedPlug() directly.

Especially since the concept of "implicit on classic or app installed on core and classic" is relatively new (as mentioned before, it is only avahi* and cups-control at this point), I wanted to err on the side of caution and have unit tests that explicitly showed it was working properly. If you'd prefer that I remove this conditional and we just rely on implicit* working as they should, I can see that point and can adjust.

anonymouse64 replied:
This is fine, you can leave this in, I just wanted to make sure I wasn't missing something and that this logic was just extra caution.

jdstrand replied:
UPDATE: there are actually several interfaces besides avahi* and cups that have this base decl/implicit* combination, so I added a new test for these.

jdstrand commented:
I have been communicating with @mvo5, et al that this was for 2.46, but I didn't milestone it, so I did just now.

@anonymouse64 - I believe that I addressed all you feedback, so feel free to re-review (I'm going to keep an eye on the spread tests though since they aren't all done running yet). Assuming the spread tests pass, all that is left is the extra test Samuele asked for (which I'm working on now).

jdstrand commented:
@anonymouse64 - ok, I think this is ready for your re-review (I fixed the spread issue and committed a moment ago).

@pedronis - I believe this is ready for your re-review as well (I've addressed your feedback).

jdstrand commented:
@mvo5 - the spread tests are failing due to this unrelated failure:
```
2020-08-19 04:32:33 Failed tasks: 4
    - google:ubuntu-16.04-64:tests/main/lxd:snapd_cgroup_both
    - google:ubuntu-16.04-64:tests/main/lxd:snapd_cgroup_just_inside
    - google:ubuntu-16.04-64:tests/main/lxd:snapd_cgroup_just_outside
    - google:ubuntu-16.04-64:tests/main/lxd:snapd_cgroup_neither
```

This is ready for review.

jdstrand commented:
I created https://github.com/snapcore/snapd/pull/9194 for 2.46 so removed the 2.46 milestone from this PR.

jdstrand commented:
@zyga reviewed https://github.com/snapcore/snapd/pull/9194, but it was closed stating that 2.46 will have a pull from master, so adding the 2.46 milestone back to this PR. This PR remains ready to review.

pedronis commented on interfaces/builtin/cups_control.go:
```diff
@@ -19,30 +19,159 @@
+}
+
+func (iface *cupsControlInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	// Only apply slot snippet when running as application snap on classic
```
this comment is confusing,  on core the helper is always false, which means we apply the snippet. I think the comment would be clearer as:

// On classic only apply slot snippet when...

pedronis commented on interfaces/builtin/cups_control.go:
```diff
@@ -19,30 +19,159 @@
+}
+
+func (iface *cupsControlInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Only apply slot snippet when running as application snap on classic
```
same

pedronis reviewed (approved):
lgtm, I haven't looked at the tests again/closely though

zyga reviewed (approved):
The base policy parts are really confusing and complex. I left a few remarks inline but those are non-binding. I think my head hurts now.

jdstrand commented:
Thanks for all the reviews and merge!
